### ISSUE TITLE
Add Printer abstraction

### ISSUE THREAD

**Issue by travisbrown** [circe/circe#547]
Right now we provide a `Parser` interface, but `Printer` is a case class. This is because from the beginning parsing was intended to be provided by third-party libraries, while we had our own printer. Now that circe-jackson is being more widely used for printing, it probably makes sense to give `Printer` the same treatment as `Parser`. We'd need to identify what configuration options we should expect any implementation to provide, etc.

I don't think there's any rush on this, but if we're doing it it should happen before 1.0, and the sooner the better.

/cc @vkostyukov

**Comment by herberteuler** [comment]
I don't know if this is related, but would it be appropriate to provide the ability to write to any `Appendable` for `Printer`? `printJsonAtDepth` is already capable of doing this but it is private. This is useful in some cases, e.g. to print the value of a Json directly into a Netty's `ByteBuf`.

**Comment by voidcontext** [comment]
@travisbrown I started looking into the code and I already have some ideas, but there's something that I'd like to double check: do we want to keep full backward compatibility during the refactoring?

**Comment by travisbrown** [comment]
@voidcontext No, it's not necessary. My goal for 0.10 is to aim for source compatibility, but I'd rather break stuff before 1.0 than get stuck with an API we're not completely happy with.
